FITS_rec._scale_back(): Incorrectly checks heapsize for TFORMn='Q'

BillCleveland-USRA opened the pull request:
### Description

Package: astropy.io.fits
Module: fitsrec.py

If a table is created with a column using TFORMn="Q(...)" and the heapsize becomes greater than 2**31, a ValueError that recommends the user to use the "Q" format is incorrectly raised.

```python
The problem originates in line 1133 of fitsrec.py within the "main" branch:

                if heapsize >= 2**31:
                    raise ValueError(
                        "The heapsize limit for 'P' format has been reached. "
                        "Please consider using the 'Q' format for your file."
                    )
```

Since _FormatQ is a subclass of _FormatP, the condition at line 1110 if isinstance(recformat, _FormatP): is True for both types. The above heapsize check is performed when 1110 condition is True.

The fix is to modify line 1133 of fitsrec.py to:

```python
              if type(recformat) == _FormatP and heapsize >= 2**31:
                    raise ValueError(
                        "The heapsize limit for 'P' format has been reached. "
                        "Please consider using the 'Q' format for your file."
                    )
```

This causes the heapsize check to be only performed when recformat's type is _FormatP.


<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->


<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #14808

pllim commented:
Thank you! Since this is a bug fix, it would also need a change log in https://github.com/astropy/astropy/tree/main/docs/changes/io.fits , following instructions laid out by https://github.com/astropy/astropy/blob/main/docs/changes/README.rst .

Usually for bug fix, we also recommend adding a regression test though with only enough data to reproduce the problem. Not sure if it is easy to do in your case or not, so I'll leave that to the FITS maintainers to decide.

cmarmo commented:
Hello!
If I may step in ... I plead guilty for introducing the bug in the code... thanks @BillCleveland-USRA for fixing it!
The related test is here:
https://github.com/astropy/astropy/blob/cc73b24619ce37f2af26a0140bbdda8015ac8265/astropy/io/fits/tests/test_table.py#L3207-L3231

I think it would be nice to check the fact that for Q format the error is not raised.

pllim commented:
@cmarmo , just for the record, introduced in https://github.com/astropy/astropy/pull/13429 ? In that case, existing milestone still stands. Thanks for the clarification!

@BillCleveland-USRA , at some point, the commits will need to be squashed, but maybe after adding that test? Thanks!

BillCleveland-USRA commented:
@cmarmo When I modified 5.2.2 and used it, I no longer experienced the error while processing large data files.

I'm not sure I have time to come up with a regression test for it, but I provided the following instructions on how to test this with our software:

Download and install the latest FTOOLS from the HEASARC
https://heasarc.gsfc.nasa.gov/lheasoft/download.html

IXPE python tools uses Astropy to read and process FITS files.

Download the data files
https://heasarc.gsfc.nasa.gov/FTP/ixpe/data/obs/02/02001099/event_l1/ixpe02001001_det1_evt1_v03.fits.gz
https://heasarc.gsfc.nasa.gov/FTP/ixpe/data/obs/02/02001099/hk/ixpe02001001_det1_att_v01.fits.gz

Run our IXPE utility:
```bash
ixpedet2j2000.py sc=y clobber=y \
  infile=ixpe02001001_det1_evt1_v03.fits.gz \
  attitude=ixpe02001001_det1_att_v01.fits.gz \
  outfile=det2j2000_out_1.fits
```

cmarmo commented:
> I'm not sure I have time to come up with a regression test for it,

You might want to reuse my test, modifying the type of the column
```py
import numpy as np
from astropy.io import fits

# a matrix with variable length array elements is created
nelem = 2**28
matrix = np.zeros(1, dtype=np.object_)
matrix[0] = np.arange(0.0, float(nelem + 1))

col = fits.Column(name="MATRIX", format=f"QD({nelem})", unit="", array=matrix)

t = fits.BinTableHDU.from_columns([col])
t.name = "MATRIX"

with pytest.raises(
    ValueError, match="Please consider using the 'Q' format for your file."
):
t.writeto("matrix.fits", overwrite=True)
```
this code throws the error and it shouldn't.

BillCleveland-USRA commented:
Sure thing. I'll push an update later today.

BillCleveland-USRA commented:
I added the regression test, and confirmed its validity by:

- Removing the fix, running the test, and confirming that it fails due to the ValueError being raised.
- Reinserting the fix, rerunning the test, and  confirming it now passes thanks to the ValueError not being raised.

I believe I've done all the required tasks:

- Created an issue
- Created a branch named for the issue
- Modified the code to fix the issue
- Added the regression test to verify the issue has been fixed
- Created a pull request that fixes the issue
- Added a "what's new" file for the pull request
- Squashed all the commits into a single one.

Thanks for your patience while I learn the process. I'm ready to contribute more in the future.

saimn reviewed (approved):
Looks good, thanks @BillCleveland-USRA for the PR (and @cmarmo for the review :)).
